File: src/ait/cli/commands/stash.py
from __future__ import annotations

from ..shared import export_app_namespace

export_app_namespace(globals())


@stash_app.command(
    "save",
    help="Save temporary local-only WIP into a stash snapshot without advancing the current line head.",
    short_help="Save temporary local WIP.",
)
def stash_save(
    message: Optional[str] = typer.Option(None, "--message", help="Optional stash note for later inspection."),
    keep_workspace: bool = typer.Option(
        False,
        "--keep-workspace",
        help="Keep the current workspace content after saving instead of restoring the current line head.",
    ),
    json_output: bool = typer.Option(False, "--json"),
):
    ctx = _ctx()
    try:
        data = _run_locked_task_bound_authoring_command(
            ctx,
            "stash save",
            lambda: create_stash(ctx, message, keep_workspace=keep_workspace),
        )
    except (ValueError, WorkspaceCommandBusyError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    _emit(data, json_output)


@stash_app.command(
    "list",
    help="List temporary local-only stashes without touching workspace content.",
    short_help="List temporary stashes.",
)
def stash_list(json_output: bool = typer.Option(False, "--json")):
    rows = list_stashes(_ctx())
    if json_output:
        _emit(rows, True)
        return
    table = Table(title="ait stashes")
    table.add_column("stash_id")
    table.add_column("line")
    table.add_column("snapshot_id")
    table.add_column("files")
    table.add_column("created_at")
    table.add_column("message")
    for row in rows:
        table.add_row(
            row["stash_id"],
            row["source_line_name"],
            row["snapshot_id"],
            str(row["file_count"]),
            row["created_at"],
            row.get("message") or "",
        )
    rprint(table)


@stash_app.command(
    "show",
    help="Inspect one temporary stash without restoring its workspace content.",
    short_help="Inspect one stash.",
)
def stash_show(stash_id: str, json_output: bool = typer.Option(False, "--json")):
    try:
        data = get_stash(_ctx(), stash_id)
    except KeyError as exc:
        raise typer.BadParameter(str(exc)) from exc
    _emit(data, json_output)


@stash_app.command(
    "apply",
    help="Restore workspace content from a stash without dropping the stash record.",
    short_help="Restore a stash into the workspace.",
)
def stash_apply(
    stash_id: str,
    force: bool = typer.Option(False, "--force", help="Overwrite unsaved workspace changes while applying the stash."),
    json_output: bool = typer.Option(False, "--json"),
):
    ctx = _ctx()
    try:
        data = _run_locked_task_bound_authoring_command(
            ctx,
            "stash apply",
            lambda: apply_stash(ctx, stash_id, force=force, drop=False),
        )
    except (KeyError, ValueError, IsADirectoryError, WorkspaceCommandBusyError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    _emit(data, json_output)


@stash_app.command(
    "pop",
    help="Restore workspace content from a stash and then drop the stash record.",
    short_help="Restore and drop a stash.",
)
def stash_pop(
    stash_id: str,
    force: bool = typer.Option(False, "--force", help="Overwrite unsaved workspace changes while restoring the stash."),
    json_output: bool = typer.Option(False, "--json"),
):
    ctx = _ctx()
    try:
        data = _run_locked_task_bound_authoring_command(
            ctx,
            "stash pop",
            lambda: apply_stash(ctx, stash_id, force=force, drop=True),
        )
    except (KeyError, ValueError, IsADirectoryError, WorkspaceCommandBusyError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    _emit(data, json_output)


@stash_app.command(
    "drop",
    help="Delete a stash record without restoring its workspace content.",
    short_help="Delete a stash.",
)
def stash_drop(stash_id: str, json_output: bool = typer.Option(False, "--json")):
    ctx = _ctx()
    try:
        data = _run_locked_task_bound_authoring_command(ctx, "stash drop", lambda: drop_stash(ctx, stash_id))
    except (KeyError, WorkspaceCommandBusyError) as exc:
        raise typer.BadParameter(str(exc)) from exc
    _emit(data, json_output)

File: ait-core/rust/crates/ait-cli/src/main_app/stash.rs
fn run_stash(repo: RepoRuntime, command: StashCommand) -> Result<(), String> {
    match command {
        StashCommand::Save(args) => {
            let payload = run_locked_workspace_command(&repo, "ait-cli stash save", || {
                stash_save(&repo, args.message.as_deref(), args.keep_workspace)
            })?;
            emit_stash_record("ait-cli stash save", &payload, args.json)?;
        }
        StashCommand::List(args) => {
            let payload = stash_list(&repo)?;
            if args.json {
                print_json(&payload)?;
            } else {
                let rows = payload
                    .as_array()
                    .ok_or_else(|| "stash list payload must decode to a list.".to_string())?;
                print_list(
                    rows,
                    &[
                        "stash_id",
                        "source_line_name",
                        "snapshot_id",
                        "file_count",
                        "created_at",
                        "message",
                    ],
                );
            }
        }
        StashCommand::Show(args) => {
            let payload = stash_show(&repo, &args.stash_id)?;
            emit_stash_record("ait-cli stash show", &payload, args.json)?;
        }
        StashCommand::Apply(args) => {
            let payload = run_locked_workspace_command(&repo, "ait-cli stash apply", || {
                stash_apply(&repo, &args.stash_id, args.force)
                    .map_err(|err| stash_restore_error(err, args.force))
            })?;
            emit_stash_restore("ait-cli stash apply", &payload, args.json)?;
        }
        StashCommand::Pop(args) => {
            let payload = run_locked_workspace_command(&repo, "ait-cli stash pop", || {
                stash_pop(&repo, &args.stash_id, args.force)
                    .map_err(|err| stash_restore_error(err, args.force))
            })?;
            emit_stash_restore("ait-cli stash pop", &payload, args.json)?;
        }
        StashCommand::Drop(args) => {
            let payload = run_locked_workspace_command(&repo, "ait-cli stash drop", || {
                stash_drop(&repo, &args.stash_id)
            })?;
            emit_result(
                "ait-cli stash drop",
                &payload,
                args.json,
                &["stash_id", "snapshot_id", "dropped", "snapshot_deleted"],
            )?;
        }
    }
    Ok(())
}

fn stash_restore_error(err: String, force: bool) -> String {
    if !force && err.starts_with("Workspace has unsaved changes") {
        return format!("{err}. Re-run with --force to replace those workspace changes.");
    }
    err
}

fn emit_stash_record(title: &str, payload: &JsonValue, json_output: bool) -> Result<(), String> {
    emit_result(
        title,
        payload,
        json_output,
        &[
            "stash_id",
            "source_line_name",
            "snapshot_id",
            "file_count",
            "total_bytes",
            "workspace_cleared",
            "created_at",
            "message",
        ],
    )
}

fn emit_stash_restore(title: &str, payload: &JsonValue, json_output: bool) -> Result<(), String> {
    emit_result(
        title,
        payload,
        json_output,
        &[
            "stash_id",
            "source_line_name",
            "snapshot_id",
            "current_line",
            "applied",
            "dropped",
            "snapshot_deleted",
            "created_at",
            "message",
        ],
    )
}
